instrumentation/spring/spring-boot-autoconfigure/src/test/java/io/opentelemetry/instrumentation/spring/autoconfigure/aspects/WithSpanAspectTest.java

```diff
@@ -39,99 +43,191 @@ public String testWithSpanWithValue() {
     }
 
-    @WithSpan(kind = SpanKind.CLIENT)
-    public String testWithSpanWithKind() {
-      return "Span with name testWithSpanWithKind and SpanKind.CLIENT was created";
-    }
-
     @WithSpan
     public String testWithSpanWithException() throws Exception {
       throw new Exception("Test @WithSpan With Exception");
     }
-  }
 
-  @Mock private Tracer tracer;
-  @Mock private Span span;
-  @Mock private SpanBuilder spanBuilder;
+    @WithSpan(kind = CLIENT)
+    public String testWithClientSpan(Runnable nestedCall) {
+      if (nestedCall != null) {
+        nestedCall.run();
+      }
+      return "Span with name testWithClientSpan and SpanKind.CLIENT was created";
+    }
+
+    @WithSpan(kind = SpanKind.SERVER)
+    public String testWithServerSpan(Runnable nestedCall) {
+      if (nestedCall != null) {
+        nestedCall.run();
+      }
+      return "Span with name testWithServerSpan and SpanKind.SERVER was created";
+    }
+  }
 
   private WithSpanTester withSpanTester;
 
   @BeforeEach
   void setup() {
-    // TODO(anuraaga): Replace mocking with a real tracer, this is more fragile than it needs be.
-    when(tracer.spanBuilder(any())).thenReturn(spanBuilder);
-    when(spanBuilder.setSpanKind(any())).thenReturn(spanBuilder);
-    when(spanBuilder.setParent(any())).thenReturn(spanBuilder);
-    when(spanBuilder.startSpan()).thenReturn(span);
-    when(span.storeInContext(any())).thenReturn(Context.root());
-
     AspectJProxyFactory factory = new AspectJProxyFactory(new WithSpanTester());
-    WithSpanAspect aspect = new WithSpanAspect(tracer);
+    WithSpanAspect aspect = new WithSpanAspect(instrumentation.getOpenTelemetry());
     factory.addAspect(aspect);
 
     withSpanTester = factory.getProxy();
   }
 
+  @AfterAll
+  static void tearDown() {
+    GlobalOpenTelemetry.resetForTest();
+  }
+
   @Test
   @DisplayName("when method is annotated with @WithSpan should wrap method execution in a Span")
   void withSpan() throws Throwable {
-
-    withSpanTester.testWithSpan();
-
-    verify(tracer, times(1)).spanBuilder("WithSpanTester.testWithSpan");
-    verify(spanBuilder, times(1)).startSpan();
-    verify(span, times(1)).end();
+    // when
+    runUnderTrace(
+        "parent",
+        () -> {
+          withSpanTester.testWithSpan();
+          return null;
+        });
+
+    // then
+    List<List<SpanData>> traces = instrumentation.waitForTraces(1);
+    assertThat(traces)
+        .hasSize(1)
+        .hasTracesSatisfyingExactly(
+            trace ->
+                trace
+                    .hasSize(2)
+                    .hasSpansSatisfyingExactly(
+                        parentSpan -> parentSpan.hasName("parent").hasKind(INTERNAL),
+                        span ->
+                            span.hasName("WithSpanTester.testWithSpan")
+                                .hasKind(INTERNAL)
+                                // otel SDK assertions need some work before we can comfortably use
```

Comment:
Will `hasParentSpanIndex`, the index within the trace, solve it?

Reply:
Yes, I think it will.

I initially thought that we could use the same pattern as in the javaagent asserts: `hasParentSpan(trace.span(0))` but maybe just the index is better.
On a slightly related topic: what about asserting on span links? We would have to pass the trace index there too, ex. `hasLinkIndex(traceIndex, spanIndex)`